prusti-viper/src/encoder/array_encoder.rs

```diff
@@ -0,0 +1,237 @@
+// © 2021, ETH Zurich
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use rustc_middle::ty;
+use std::{
+    collections::HashMap,
+    convert::TryInto,
+};
+use crate::encoder::{
+    Encoder,
+    errors::EncodingResult,
+    builtin_encoder::BuiltinFunctionKind,
+};
+use prusti_common::{
+    vir,
+    vir_local,
+};
+
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct EncodedArrayTypes<'tcx> {
+    pub array_pred: String,
+    pub array_ty: vir::Type,
+    pub elem_ty: vir::Type,
+    pub elem_value_ty: vir::Type,
+    pub elem_ty_rs: ty::Ty<'tcx>,
+    pub array_len: usize,
+    pub lookup_pure_name: String,
```

Comment:
I wonder: can `lookup_pure` (and others) always just have a constant name? E.g. for snapshots the snapshot function is called `snap$` for any (non-primitive etc) type. Prusti later does mangling based on the types found in the signature so the function name becomes unique again.

Reply:
i had the same thought some time ago, but i think it's at least not as simple as just dropping them from the name. currently we have
```
function Array$5$i32$lookup_pure__$TY$__Array$5$i32$$int$$$int$(self: Ref, idx: Int): Int
  requires acc(Array$5$i32(self), read$())
  requires idx < 5
```
so the different `lookup_pure` instances are actually different because they mention their respective type predicate and length in their preconditions. we might be able to get rid of the preconditions and sort-of inline them as assertions whenever we use them, but at that point i'm not sure it's worth it, so i dropped the idea at the time.

Reply:
The functions would still be different, they would all just share the same name. In the example you have, everything after `__$TY$__` is not really part of the name, it is added by name mangling later.

Reply:
ah, that sounds like a plan, yes. i'll change that.